python/cudf/cudf/core/indexed_frame.py

```diff
@@ -1694,6 +1694,204 @@ def last(self, offset):
+        if frac is None:
+            n = 1 if n is None else n
+        else:
+            if frac > 1 and not replace:
+                raise ValueError(
+                    "Replace has to be set to `True` when upsampling the "
+                    "population `frac` > 1."
+                )
+            if n is not None:
+                raise ValueError(
+                    "Please enter a value for `frac` OR `n`, not both."
+                )
+            n = int(round(size * frac))
+
+        if n > 0 and size == 0:
+            raise ValueError(
+                "Cannot take a sample larger than 0 when axis is empty."
+            )
+
+        if isinstance(random_state, cp.random.RandomState):
+            lib = cp
+        elif isinstance(random_state, np.random.RandomState):
+            lib = np
+        else:
+            # Construct random state if `random_state` parameter is None or a
+            # seed. By default, cupy random state is used to sample from rows
+            # and numpy is used to sample from columns. In general, cuDF
+            # assumes the number of columns is much smaller than the number of
+            # rows, thus using numpy random states can avoid kernel launching
+            # overhead on creating a small gather map.
+            lib = cp if axis == 0 else np
+            random_state = lib.random.RandomState(seed=random_state)
+
+        # Normalize `weights` array.
+        if weights is not None:
+            if isinstance(weights, str):
+                raise NotImplementedError(
+                    "Weights specified by string is unsupported yet."
+                )
+
+            if size != len(weights):
+                raise ValueError(
+                    "Weights and axis to be sampled must be of same length."
+                )
+
+            weights = lib.asarray(weights)
+            weights = weights / weights.sum()
+
+        if axis == 0:
+            return self._sample_axis_0(
+                n, weights, replace, random_state, ignore_index
+            )
+        else:
+            if isinstance(random_state, cp.random.RandomState):
+                raise ValueError(
+                    "Sampling from `axis=1`/`columns` with cupy random state"
+                    "isn't supported."
+                )
+            return self._sample_axis_1(
+                n, weights, replace, random_state, ignore_index
+            )
+
+    def _sample_axis_0(
+        self,
+        n: int,
+        weights: Optional[ColumnLike],
+        replace: bool,
+        random_state: Union[np.random.RandomState, cp.random.RandomState],
+        ignore_index: bool,
+    ):
+        try:
+            gather_map_array = random_state.choice(
+                len(self), size=n, replace=replace, p=weights
+            )
+        except NotImplementedError as e:
+            raise NotImplementedError(
+                "Unsupported arguments for gather map generation."
+            ) from e
```

Comment:
Is this specifically for the weights parameter? The message isn't that descriptive, if we're going to catch the original NotImplementedError and rethrow a new one, we should provide more information here.

Reply:
The thing is `cupy.randomstate.choice` has several code paths that throws `NotImplementedError` and not all of them are well described with a message: https://github.com/cupy/cupy/blob/fa4993a74755d3dcd0affebe0a372c594d6a8679/cupy/random/_generator.py#L1010.

I can do a best effort to check the argument combinations here if that's desired.

Reply:
Maybe just change the message to say "Random sampling with cupy does not support these inputs" so that it's obvious that the failure is in trying to use `cupy.random.RandomState.choice`.